bionic/code_analyzer.py

```diff
@@ -0,0 +1,203 @@
+
+        # Add a prefix for obj type so we can differentiate between
+        # different types that look the same, like a list and a tuple,
+        # or an int and bytes.
+        self._add_prefix_for_type(obj)
+
+        self._keys_being_hashed.add(obj_key)
+        self._analyze(obj)
+        self._keys_being_hashed.remove(obj_key)
+
+    def _add_prefix_for_type(self, obj):
+        if isinstance(obj, bytes):
+            prefix = b"A"
+        elif isinstance(obj, bytearray):
+            prefix = b"B"
+        elif obj is None:
+            prefix = b"C"
+        elif isinstance(obj, str):
+            prefix = b"D"
+        elif isinstance(obj, int):
+            prefix = b"E"
+        elif isinstance(obj, float):
+            prefix = b"F"
+        elif isinstance(obj, bool):
+            prefix = b"G"
+        elif isinstance(obj, list):
+            prefix = b"H"
+        elif isinstance(obj, tuple):
+            prefix = b"I"
+        elif isinstance(obj, dict):
+            prefix = b"J"
+        # Add any new types here. If you change the prefix of any type,
+        # or add a new prefix, it can change the encoding of objects.
+        # Since Bionic uses encoding to detect any changes, any changes
+        # to encoding makes caching backwards incompatible.
+        #
+        # Update CACHE_SCHEMA_VERSION to update cache scheme if you
+        # change the encoding.
+        else:
+            prefix = b"Z"
+
+        self._analyzed_bytearray += prefix
+
+    def _add_results(self, obj, result):
+        self._results_by_key[get_obj_key(obj)] = result
+        self._analyzed_bytearray += result
+
+    def _analyze(self, obj):
+        """
+        Contains the logic that analyzes the objects and converts them
+        into bytes. The converted bytes are added to the result object
+        instead of returning them.
+        """
+        if isinstance(obj, bytes) or isinstance(obj, bytearray):
+            self._add_results(obj, obj)
+
+        elif obj is None or isinstance(obj, (int, float, str, bool)):
+            json_bytes = json.dumps(obj).encode()
+            self._add_results(obj, json_bytes)
+
+        elif isinstance(obj, (list, tuple)):
+            for elem in obj:
+                self.analyze(elem)
+
+        elif isinstance(obj, dict):
+            for key, elem in obj.items():
+                self.analyze(key)
+                self.analyze(elem)
+
+        elif inspect.isroutine(obj):
+            if obj.__defaults__ is not None:
+                for d in obj.__defaults__:
+                    self.analyze(d)
+
+            self._analyze_code(obj.__code__)
+
+        elif inspect.iscode(obj):
+            self._analyze_code(obj)
+
+        elif inspect.isclass(obj):
+            # TODO: Add support to analyze class objects.
+            self.analyze(obj.__name__)
+
+        else:
+            message = oneline(
+                f"""
+                Found complex object {obj!r} of type {type(obj)!r}
```

Comment:
I don't think there are any constants that we don't handle, but I left it here just in case there is one that I missed. It will be easy for users to notice a problem and contact us.